Plugin client must print an error message in case of create error

Recent changes to formatting of the disks on the guest side may have introduced this error. When create hits an error condition we get this,

root@photon-jyLxAJ3OK [ ~ ]# docker volume create -d vmdk --name v2
Error response from daemon: create v2: Post http://%2Frun%2Fdocker%2Fplugins%2Fvmdk.sock/VolumeDriver.Create: http: ContentLength=24 with Body length 0

Ok, this is a repro,

1. Create volume v6
root@photon-jyLxAJ3OK [ /sys ]# docker volume create -d vmdk --name v6

Error response from daemon: create v6: Post
http://%2Frun%2Fdocker%2Fplugins%2Fvmdk.sock/VolumeDriver.Create:
http: ContentLength=24 with Body length 0


2. Client logs
2016-10-07 09:01:05.127970653 +0000 UTC [INFO] Attaching volume and
creating filesystem fstype=ext4 name=v6
2016-10-07 09:01:05.740203226 +0000 UTC [WARNING] Get device path failed
for unit# %s @ PCI slot %s: 1192Error="open /sys/bus/pci/slots/192/address:
no such file or directory"
2016-10-07 09:01:05.74026955 +0000 UTC [ERROR] Could not find attached
device, removing the volume name=v6 error="Device not found"

3. Server logs
10/07/16 09:01:03 117321 [P1] [INFO   ] executeRequest 'get' completed with
ret={u'Error': 'Volume v6 not found (file:
/vmfs/volumes/Storage1/dockvols/v6.vmdk)'}
10/07/16 09:01:03 117321 [P1] [INFO   ] *** createVMDK:
/vmfs/volumes/Storage1/dockvols/v6.vmdk opts = {u'fstype': u'ext4'}
10/07/16 09:01:05 117321 [P1] [INFO   ] executeRequest 'create' completed
with ret=None
10/07/16 09:01:05 117321 [P1] [INFO   ] *** attachVMDK:
/vmfs/volumes/Storage1/dockvols/v6.vmdk to P1 VM uuid =
564dabc0-da80-2e80-1174-d92d69585cc4
10/07/16 09:01:05 117321 [P1] [INFO   ] Attaching
/vmfs/volumes/Storage1/dockvols/v6.vmdk as independent_persistent
10/07/16 09:01:06 117321 [P1] [INFO   ] Disk
/vmfs/volumes/Storage1/dockvols/v6.vmdk successfully attached. controller
pci_slot_number=192, disk_slot=1
10/07/16 09:01:06 117321 [P1] [INFO   ] executeRequest 'attach' completed
with ret={'ControllerPciSlotNumber': '192', 'Unit': '1'}
10/07/16 09:01:06 117321 [P1] [INFO   ] *** detachVMDK:
/vmfs/volumes/Storage1/dockvols/v6.vmdk from P1 VM uuid =
564dabc0-da80-2e80-1174-d92d69585cc4
10/07/16 09:01:06 117321 [P1] [INFO   ] Disk detached
/vmfs/volumes/Storage1/dockvols/v6.vmdk
10/07/16 09:01:06 117321 [P1] [INFO   ] executeRequest 'detach' completed
with ret=None
10/07/16 09:01:06 117321 [P1] [INFO   ] *** removeVMDK:
/vmfs/volumes/Storage1/dockvols/v6.vmdk
10/07/16 09:01:07 117321 [P1] [INFO   ] executeRequest 'remove' completed
with ret=None

The guest is running docker 1.12 with API version 1.24. Seems like error
response to the docker daemon may be an issue or a docker bug itself.


On Fri, Oct 7, 2016 at 12:33 PM, Bruno Moura <notifications@github.com>
wrote:

> These are the server logs, can you add the plugin logs?
>
> thanks!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/vmware/docker-volume-vsphere/issues/616#issuecomment-252168052>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/APHseBZqgkwIPfRJOsiEVyQFUXJ-gsffks5qxe7HgaJpZM4KQsdo>
> .
>

--- a/vmdk_plugin/plugin.go
+++ b/vmdk_plugin/plugin.go
@@ -235,66 +235,61 @@ func (d *vmdkDriver) Create(r volume.Request) volume.Response {
 		return volume.Response{Err: msg + validfs}
 	}
 
-	err := d.ops.Create(r.Name, r.Options)
-	if err != nil {
-		log.WithFields(log.Fields{"name": r.Name, "error": err}).Error("Create volume failed ")
-		return volume.Response{Err: err.Error()}
+	errCreate := d.ops.Create(r.Name, r.Options)
+	if errCreate != nil {
+		log.WithFields(log.Fields{"name": r.Name, "error": errCreate}).Error("Create volume failed ")
+		return volume.Response{Err: errCreate.Error()}
 	}
 
 	// Handle filesystem creation
 	log.WithFields(log.Fields{"name": r.Name,
 		"fstype": r.Options["fstype"]}).Info("Attaching volume and creating filesystem ")
 
-	dev, err := d.ops.Attach(r.Name, nil)
-	if err != nil {
+	dev, errAttach := d.ops.Attach(r.Name, nil)
+	if errAttach != nil {
 		log.WithFields(log.Fields{"name": r.Name,
-			"error": err}).Error("Attach volume failed, removing the volume ")
-		err = d.ops.Remove(r.Name, nil)
-		if err != nil {
-			log.WithFields(log.Fields{"name": r.Name, "error": err}).Error("Remove volume failed ")
-			return volume.Response{Err: err.Error()}
+			"error": errAttach}).Error("Attach volume failed, removing the volume ")
+		errRemove := d.ops.Remove(r.Name, nil)
+		if errRemove != nil {
+			log.WithFields(log.Fields{"name": r.Name, "error": errRemove}).Warning("Remove volume failed ")
 		}
-		return volume.Response{Err: err.Error()}
+		return volume.Response{Err: errAttach.Error()}
 	}
 
-	device, err := fs.GetDevicePath(dev)
-	if err != nil {
+	device, errGetDevicePath := fs.GetDevicePath(dev)
+	if errGetDevicePath != nil {
 		log.WithFields(log.Fields{"name": r.Name,
-			"error": err}).Error("Could not find attached device, removing the volume ")
-		err = d.ops.Detach(r.Name, nil)
-		if err != nil {
-			log.WithFields(log.Fields{"name": r.Name, "error": err}).Error("Detach volume failed ")
-			return volume.Response{Err: err.Error()}
+			"error": errGetDevicePath}).Error("Could not find attached device, removing the volume ")
+		errDetach := d.ops.Detach(r.Name, nil)
+		if errDetach != nil {
+			log.WithFields(log.Fields{"name": r.Name, "error": errDetach}).Warning("Detach volume failed ")
 		}
-		err = d.ops.Remove(r.Name, nil)
-		if err != nil {
-			log.WithFields(log.Fields{"name": r.Name, "error": err}).Error("Remove volume failed ")
-			return volume.Response{Err: err.Error()}
+		errRemove := d.ops.Remove(r.Name, nil)
+		if errRemove != nil {
+			log.WithFields(log.Fields{"name": r.Name, "error": errRemove}).Warning("Remove volume failed ")
 		}
-		return volume.Response{Err: err.Error()}
+		return volume.Response{Err: errGetDevicePath.Error()}
 	}
 
-	err = fs.Mkfs(mkfscmd, r.Name, device)
-	if err != nil {
+	errMkfs := fs.Mkfs(mkfscmd, r.Name, device)
+	if errMkfs != nil {
 		log.WithFields(log.Fields{"name": r.Name,
-			"error": err}).Error("Create filesystem failed, removing the volume ")
-		err = d.ops.Detach(r.Name, nil)
-		if err != nil {
-			log.WithFields(log.Fields{"name": r.Name, "error": err}).Error("Detach volume failed ")
-			return volume.Response{Err: err.Error()}
+			"error": errMkfs}).Error("Create filesystem failed, removing the volume ")
+		errDetach := d.ops.Detach(r.Name, nil)
+		if errDetach != nil {
+			log.WithFields(log.Fields{"name": r.Name, "error": errDetach}).Warning("Detach volume failed ")
 		}
-		err = d.ops.Remove(r.Name, nil)
-		if err != nil {
-			log.WithFields(log.Fields{"name": r.Name, "error": err}).Error("Remove volume failed ")
-			return volume.Response{Err: err.Error()}
+		errRemove := d.ops.Remove(r.Name, nil)
+		if errRemove != nil {
+			log.WithFields(log.Fields{"name": r.Name, "error": errRemove}).Warning("Remove volume failed ")
 		}
-		return volume.Response{Err: err.Error()}
+		return volume.Response{Err: errMkfs.Error()}
 	}
 
-	err = d.ops.Detach(r.Name, nil)
-	if err != nil {
-		log.WithFields(log.Fields{"name": r.Name, "error": err}).Error("Detach volume failed ")
-		return volume.Response{Err: err.Error()}
+	errDetach := d.ops.Detach(r.Name, nil)
+	if errDetach != nil {
+		log.WithFields(log.Fields{"name": r.Name, "error": errDetach}).Error("Detach volume failed ")
+		return volume.Response{Err: errDetach.Error()}
 	}
 
 	log.WithFields(log.Fields{"name": r.Name,

